### PR TITLE
Reduce memory usage of streaming fast-lossless code-path.

### PR DESCRIPTION
**Pull request by szabadka**
Memory usage for a 45MP image went from 466 MB to 6.7 MB.

cjxl -d 0.0 -e 1 --streaming_input --streaming_output

BEFORE:
Compressed to 79443.2 kB (13.986 bpp).
8256 x 5504, 125.388 MP/s [125.39, 125.39], 1 reps, 8 threads.
AFTER:
Compressed to 79444.1 kB (13.986 bpp).
8256 x 5504, 105.129 MP/s [105.13, 105.13], 1 reps, 8 threads.

cjxl  -d 0.0 -e 1

BEFORE:
Compressed to 79443.2 kB (13.986 bpp).
8256 x 5504, 229.300 MP/s [229.30, 229.30], 1 reps, 8 threads.
AFTER:
Compressed to 79443.2 kB (13.986 bpp).
8256 x 5504, 230.898 MP/s [230.90, 230.90], 1 reps, 8 threads.

### CODE REVIEW

File: lib/jxl/decode_test.cc [hunk 1]
```diff
@@ -5024,6 +5024,8 @@ bool BoxTypeEquals(const std::string& type_string, JxlBoxType type) {
 }
 }  // namespace
 
+#if 0
+// TODO(szabadka) Fix the underlying test file.
 TEST(DecodeTest, ExtentedBoxSizeTest) {
```
Comment (veluca93) [inline lib/jxl/decode_test.cc:5029]: DISABLED_ instead of #if-0

File: lib/extras/packed_image.h [hunk 2]
```diff
@@ -268,6 +272,7 @@ class ChunkedPackedFrame {
   void* opaque_;
   ReadLine read_line_;
   std::set<void*> buffers_;
+  std::mutex* mtx_;
```
Comment (veluca93) [inline lib/extras/packed_image.h:275]: I don't see why this should be an externally-owned mutex...

File: lib/jxl/decode.cc [hunk 3]
```diff
@@ -1595,14 +1595,14 @@ static JxlDecoderStatus ParseBoxHeader(const uint8_t* in, size_t size,
   // Box size, including this header itself.
   *box_size = LoadBE32(in + pos);
   pos += 4;
+  memcpy(type, in + pos, 4);
+  pos += 4;
   if (*box_size == 1) {
     *header_size = 16;
-    if (OutOfBounds(pos, 12, size)) return JXL_DEC_NEED_MORE_INPUT;
+    if (OutOfBounds(pos, 8, size)) return JXL_DEC_NEED_MORE_INPUT;
```
Comment (veluca93) [inline lib/jxl/decode.cc:1602]: Can we split this to a different commit? and possibly even PR

File: lib/jxl/enc_fast_lossless.h [hunk 4]
```diff
@@ -46,9 +47,15 @@ struct JxlFastLosslessFrameState;
 // Returned JxlFastLosslessFrameState must be freed by calling
 // JxlFastLosslessFreeFrameState.
 JxlFastLosslessFrameState* JxlFastLosslessPrepareFrame(
-    const unsigned char* rgba, size_t width, size_t row_stride, size_t height,
-    size_t nb_chans, size_t bitdepth, int big_endian, int effort,
-    void* runner_opaque, FJxlParallelRunner runner);
+    JxlChunkedFrameInputSource input, size_t width, size_t height,
+    size_t nb_chans, size_t bitdepth, int big_endian, int effort);
+
+class JxlEncoderOutputProcessorWrapper;
```
Comment (veluca93) [inline lib/jxl/enc_fast_lossless.h:53]: Can you remove these if FJXL_STANDALONE is 1? also from the cc file

File: lib/jxl/encode_internal.h [hunk 5]
```diff
@@ -140,6 +140,95 @@ constexpr std::array<unsigned char, 32> kContainerHeader = {
 constexpr std::array<unsigned char, 8> kLevelBoxHeader = {0,   0,   0,   0x9,
                                                           'j', 'x', 'l', 'l'};
 
+class JxlEncoderChunkedFrameAdapter {
```
Comment (veluca93) [inline lib/jxl/encode_internal.h:143]: A comment on what this is for wouldn't hurt :-)

File: lib/jxl/encode.cc [hunk 6]
```diff
@@ -870,12 +881,25 @@ jxl::Status JxlEncoderStruct::ProcessOneEnqueuedInput() {
       timecode = 0;
     }
 
-    bool last_frame = frames_closed && !num_queued_frames;
+    const bool last_frame = frames_closed && !num_queued_frames;
 
-    jxl::BitWriter writer;
+    size_t box_header_size = jxl::kLargeBoxHeaderSize;
+    bool use_large_box = true;
```
Comment (veluca93) [inline lib/jxl/encode.cc:887]: TODO about making this conditional?

File: lib/jxl/enc_fast_lossless.cc [hunk 7]
```diff
@@ -3657,80 +3748,188 @@ JxlFastLosslessFrameState* LLEnc(const unsigned char* rgba, size_t width,
...
     }
-  };
-
-  runner(
-      runner_opaque, &run_one,
-      +[](void* r, size_t i) { (*reinterpret_cast<decltype(&run_one)>(r))(i); },
-      num_groups_x * num_groups_y);
+    return;
+  }
+  constexpr size_t kMaxLocalGroups = 16;
```
Comment (veluca93) [inline lib/jxl/enc_fast_lossless.cc:3805]: What's this?

File: lib/jxl/enc_fast_lossless.cc [hunk 8]
```diff
@@ -3657,80 +3748,188 @@ JxlFastLosslessFrameState* LLEnc(const unsigned char* rgba, size_t width,
...
-                            lookup.data(), nb_chans, gd[0]);
+template <typename BitDepth>
+void LLProcess(JxlFastLosslessFrameState* frame_state, bool is_last,
+               BitDepth bitdepth, void* runner_opaque,
+               FJxlParallelRunner runner,
+               JxlEncoderOutputProcessorWrapper* output_processor) {
+  if (frame_state->process_done) {
+    JxlFastLosslessPrepareHeader(frame_state, /*add_image_header=*/0, is_last);
+    if (output_processor) {
+      JxlFastLosslessOutputFrame(frame_state, output_processor);
```
Comment (veluca93) [inline lib/jxl/enc_fast_lossless.cc:3801]: In principle, one could start writing output even before all groups are done (i.e. one could flush out a prefix of the groups).

File: lib/jxl/enc_fast_lossless.cc [hunk 9]
```diff
@@ -3657,80 +3748,188 @@ JxlFastLosslessFrameState* LLEnc(const unsigned char* rgba, size_t width,
...
+      size_t num_dc_groups =
+          frame_state->num_dc_groups_x * frame_state->num_dc_groups_y;
+      size_t group_id = onegroup ? 0 : (2 + num_dc_groups + g);
+      size_t xs = std::min<size_t>(frame_state->width - xg * 256, 256);
+      size_t ys = std::min<size_t>(frame_state->height - yg * 256, 256);
+      size_t x0 = xg * 256;
+      size_t y0 = yg * 256;
+      size_t stride;
+      JxlChunkedFrameInputSource input = frame_state->input;
+      const void* buffer = input.get_color_channel_data_at(input.opaque, x0, y0,
```
Comment (veluca93) [inline lib/jxl/enc_fast_lossless.cc:3835]: At some point, we probably want to add a RAII wrapper around this.

File: lib/jxl/enc_fast_lossless.h (hunk 4 above)
Comment (veluca93) [reply lib/jxl/enc_fast_lossless.h:53]: In general, could you test that the scripts in tools/fast_lossless still work?

File: lib/jxl/enc_fast_lossless.cc [hunk 10]
```diff
@@ -3516,30 +3594,37 @@ bool detect_palette(const unsigned char* r, size_t width,
 }
 
 template <typename BitDepth>
-JxlFastLosslessFrameState* LLEnc(const unsigned char* rgba, size_t width,
-                                 size_t stride, size_t height,
-                                 BitDepth bitdepth, size_t nb_chans,
-                                 bool big_endian, int effort,
-                                 void* runner_opaque,
-                                 FJxlParallelRunner runner) {
+JxlFastLosslessFrameState* LLPrepare(JxlChunkedFrameInputSource input,
```
Comment (veluca93) [inline lib/jxl/enc_fast_lossless.cc:3597]: Is there a way to know at this point whether the *input* is streamed?

File: lib/jxl/decode_test.cc (hunk 1 above)
Comment (szabadka) [reply lib/jxl/decode_test.cc]: Done. Also sent a separate PR.

File: lib/extras/packed_image.h (hunk 2 above)
Comment (szabadka) [reply lib/extras/packed_image.h:275]: First I added mutex here but then in complained for the place where we emplace_back a chunked frame.

File: lib/jxl/decode.cc (hunk 3 above)
Comment (szabadka) [reply lib/jxl/decode.cc:1602]: Sent separate PR.

File: lib/jxl/enc_fast_lossless.h (hunk 4 above)
Comment (szabadka) [reply lib/jxl/enc_fast_lossless.h:53]: They are quite broken currently, starting with fatal error: 'jxl/encode.h' file not found

File: lib/jxl/encode_internal.h (hunk 5 above)
Comment (szabadka) [reply lib/jxl/encode_internal.h:145]: Done.

File: lib/jxl/encode.cc (hunk 6 above)
Comment (szabadka) [reply lib/jxl/encode.cc:889]: Done.

File: lib/jxl/enc_fast_lossless.cc (hunk 7 above)
Comment (szabadka) [reply lib/jxl/enc_fast_lossless.cc:3811]: Added comment + TODO

File: lib/jxl/enc_fast_lossless.cc (hunk 8 above)
Comment (szabadka) [reply lib/jxl/enc_fast_lossless.cc:3804]: This code-path is for the non-streaming input case where all the frames were processed already and we just generate the output.

File: lib/jxl/enc_fast_lossless.cc (hunk 9 above)
Comment (szabadka) [reply lib/jxl/enc_fast_lossless.cc:3841]: Added comment.

File: lib/jxl/enc_fast_lossless.cc (hunk 10 above)
Comment (szabadka) [reply lib/jxl/enc_fast_lossless.cc:3599]: The caller should know, I added a TODO for plumbing that it.

File: lib/extras/packed_image.h (hunk 2 above)
Comment (veluca93) [reply lib/extras/packed_image.h:275]: Ah, because it makes the struct not movable...

I'd perhaps make this a unique_ptr<mutex> instead, that feels safer

File: lib/jxl/enc_fast_lossless.cc (hunk 8 above)
Comment (veluca93) [reply lib/jxl/enc_fast_lossless.cc:3804]: Ah, I must have missed that then :)

File: lib/jxl/enc_fast_lossless.h (hunk 4 above)
Comment (szabadka) [reply lib/jxl/enc_fast_lossless.h:53]: no, I don't think so, I will open an issue for this

Comment (szabadka) [reply lib/jxl/enc_fast_lossless.h:53]: Opened https://github.com/libjxl/libjxl/issues/2959

File: lib/extras/packed_image.h (hunk 2 above)
Comment (szabadka) [reply lib/extras/packed_image.h]: Done.
